lgc/state/PalMetadata.cpp

```diff
@@ -520,33 +543,63 @@ void PalMetadata::getVertexFetchInfo(SmallVectorImpl<VertexFetchInfo> &fetches)
     unsigned location = fetchNode[0].getUInt();
     unsigned component = fetchNode[1].getUInt();
     StringRef tyName = fetchNode[2].getString();
-    unsigned vecLength = 0;
-    if (tyName[0] == 'v') {
-      tyName = tyName.drop_front();
-      tyName.consumeInteger(10, vecLength);
-    }
-    Type *ty = nullptr;
-    if (tyName == "i8")
-      ty = Type::getInt8Ty(m_pipelineState->getContext());
-    else if (tyName == "i16")
-      ty = Type::getInt16Ty(m_pipelineState->getContext());
-    else if (tyName == "i32")
-      ty = Type::getInt32Ty(m_pipelineState->getContext());
-    else if (tyName == "i64")
-      ty = Type::getInt64Ty(m_pipelineState->getContext());
-    else if (tyName == "f16")
-      ty = Type::getHalfTy(m_pipelineState->getContext());
-    else if (tyName == "f32")
-      ty = Type::getFloatTy(m_pipelineState->getContext());
-    else if (tyName == "f64")
-      ty = Type::getDoubleTy(m_pipelineState->getContext());
-    if (vecLength != 0)
-      ty = FixedVectorType::get(ty, vecLength);
+    Type *ty = getLlvmType(tyName);
     fetches.push_back({location, component, ty});
   }
   m_pipelineNode.erase(m_document->getNode(PipelineMetadataKey::VertexInputs));
 }
 
+// =====================================================================================================================
+// Store the color export information in PAL metadata for an exportless fragment shader with shader compilation.
+//
+// @param info : Array of ColorExportInfo structs
+void PalMetadata::addColorExportInfo(ArrayRef<ColorExportInfo> exports) {
+  // Each color export is an array containing {location, original location,type}.
+  // .colorExports is an array containing the color exports.
+  m_colorExports = m_pipelineNode[PipelineMetadataKey::ColorExports].getArray(true);
+  for (const ColorExportInfo &exportInfo : exports) {
+    msgpack::ArrayDocNode fetchNode = m_document->getArrayNode();
+    fetchNode.push_back(m_document->getNode(exportInfo.hwColorTarget));
+    fetchNode.push_back(m_document->getNode(exportInfo.location));
+    fetchNode.push_back(m_document->getNode(exportInfo.isSigned));
+    fetchNode.push_back(m_document->getNode(getTypeName(exportInfo.ty), /*copy=*/true));
+    m_colorExports.push_back(fetchNode);
+  }
+}
+
+// =====================================================================================================================
+// Get the count of color exports needed by the fragement shader.
+unsigned PalMetadata::getColorExportCount() {
+  if (m_colorExports.isEmpty())
+    return 0;
+  return m_colorExports.size();
```

Comment:
This should be simplified.

Reply:
I don't think we can just check the size.  If m_colorExports has not been assigned to an array node yet, then you cannot ask for the size.  Before the check the size, we have to check if the node type is "empty", which is different change checking if the array is empty.

Reply:
I'll change the type for m_colorExports to a DocNode instead of an ArrayDocNode to make this more explicit.